src/main/java/seedu/address/logic/commands/FilterTaskCommand.java

```diff
@@ -0,0 +1,178 @@
+import seedu.address.commons.util.CollectionUtil;
+import seedu.address.model.Model;
+import seedu.address.model.task.TaskCategory;
+import seedu.address.model.task.TaskCategoryAndDeadlinePredicate;
+import seedu.address.model.task.TaskContainsCategoryPredicate;
+import seedu.address.model.task.TaskDate;
+import seedu.address.model.task.TaskDeadlineBeforeDatePredicate;
+
+/**
+ * Filters and lists all tasks in address book with category that matches the argument category
+ * and / or with deadline that matches the argument date.
+ */
+public class FilterTaskCommand extends Command {
+
+    public static final String COMMAND_WORD = "filter";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Filters all tasks based on the user input "
+            + "and displays them as a list with index numbers.\n"
+            + "Parameters: "
+            + "[" + PREFIX_CATEGORY + "CATEGORY (database/frontend/backend/uiux/presentation/others)] "
+            + "[" + PREFIX_DATE + "DATE (YYYY-MM-DD)]\n"
+            + "Requirement: At least one parameter to filter with must be provided.\n"
+            + "Example: " + COMMAND_WORD + " "
+            + PREFIX_CATEGORY + "backend "
+            + PREFIX_DATE + "2022-12-12";
+
+    private final FilterTaskDescriptor filterTaskDescriptor;
+    private final TaskCategoryAndDeadlinePredicate both;
+    private final TaskContainsCategoryPredicate category;
+    private final TaskDeadlineBeforeDatePredicate date;
+    private final boolean hasCategory;
+    private final boolean hasDate;
+
+    /**
+     * @param filterTaskDescriptor details to filter the list of tasks with
+     */
+    public FilterTaskCommand(FilterTaskDescriptor filterTaskDescriptor) {
+        requireNonNull(filterTaskDescriptor);
+        this.filterTaskDescriptor = new FilterTaskDescriptor(filterTaskDescriptor);
+
+        if (filterTaskDescriptor.getCategory().isPresent() && filterTaskDescriptor.getDate().isPresent()) {
+
+            both = new TaskCategoryAndDeadlinePredicate(
+                    filterTaskDescriptor.getCategory().get(), filterTaskDescriptor.getDate().get());
+            category = null;
+            date = null;
+
+            hasCategory = true;
+            hasDate = true;
+
+        } else if (filterTaskDescriptor.getCategory().isPresent() && filterTaskDescriptor.getDate().isEmpty()) {
+
+            both = null;
+            category = new TaskContainsCategoryPredicate(filterTaskDescriptor.getCategory().get());
+            date = null;
+
+            hasCategory = true;
+            hasDate = false;
+
+        } else {
+
+            both = null;
+            category = null;
+            date = new TaskDeadlineBeforeDatePredicate(filterTaskDescriptor.getDate().get());
+
+            hasCategory = false;
+            hasDate = true;
+
+        }
+    }
+
+    @Override
+    public CommandResult execute(Model model) {
+        requireNonNull(model);
+
+        if (hasCategory && hasDate) {
```

Comment:
I also think that maybe having one string variable instead of two boolean variables is better. What do u think?

Reply:
Resolved with the updated `TaskCategoryAndDeadlinePredicate` class!